### PR TITLE
100% coverage for contrib.appengine

### PR DESCRIPTION
**Pull request by theacodes**
@dhermes this module and its existing tests are really quite messy, so I'm sorry if these tests seem a bit off.

Towards #212

### CODE REVIEW

**Comment by jonparrott** [comment]
@dhermes friendly ping on this one.

File: tests/contrib/test_appengine.py [hunk 1]
```diff
@@ -39,14 +39,20 @@
...
 from oauth2client.clientsecrets import _loadfile
+from oauth2client.clientsecrets import TYPE_WEB
 from oauth2client.clientsecrets import InvalidClientSecretsError
 from oauth2client.contrib.appengine import AppAssertionCredentials
 from oauth2client.contrib.appengine import CredentialsModel
 from oauth2client.contrib.appengine import CredentialsNDBModel
+from oauth2client.contrib.appengine import CredentialsProperty
 from oauth2client.contrib.appengine import FlowProperty
+from oauth2client.contrib.appengine import (
+    InvalidClientSecretsError as AppEngineInvalidClientSecretsError)
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:52]: Oy

File: tests/contrib/test_appengine.py [hunk 2]
```diff
@@ -308,6 +318,74 @@ def test_flow_get_put(self):
...
+        self.assertRaises(
+            db.BadValueError,
+            FlowProperty().validate, 42)
+
+
+class TestCredentialsModel(db.Model):
+    credentials = CredentialsProperty()
+
+
+class CredentialsPropertyTest(unittest2.TestCase):
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:335]: Didn't this already exist?

File: tests/contrib/test_appengine.py [hunk 3]
```diff
@@ -339,6 +417,27 @@ def setUp(self):
     def tearDown(self):
         self.testbed.deactivate()
 
+    def test_bad_ctor(self):
+        self.assertRaises(
+            ValueError, StorageByKeyName, CredentialsModel, None, None)
```
Comment (dhermes) [inline tests/contrib/test_appengine.py]: At this point (once using `unittest2`) isn't it clearer to say

``` python
with self.assertRaises(ValueError):
    StorageByKeyName(CredentialsModel, None, None)
```

File: tests/contrib/test_appengine.py [hunk 4]
```diff
@@ -522,7 +621,7 @@ def get(self):
                 parent.had_credentials = True
                 parent.found_credentials = decorator.credentials
                 if parent.should_raise:
-                    raise Exception('')
+                    raise parent.should_raise
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:624]: Teach me! What is `should_raise`?

File: tests/contrib/test_appengine.py [hunk 5]
```diff
@@ -557,6 +656,24 @@ def tearDown(self):
         self.testbed.deactivate()
         httplib2.Http = self.httplib2_orig
 
+    def test_in_error(self):
+        # Note: as far as I can tell this branch is never reached. _in_error
```
Comment (dhermes) [inline tests/contrib/test_appengine.py]: I usually say `NOTE`, also `I` isn't very helpful to use in a comment like this.

Also it seems like this note should be an issue filed and the absence of the value being set should be addressed.

File: tests/contrib/test_appengine.py [hunk 6]
```diff
@@ -610,11 +727,21 @@ def test_required(self):
         self.assertEqual(None, self.decorator.credentials)
 
         # Raising an exception still clears the Credentials.
-        self.should_raise = True
+        self.should_raise = Exception('')
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:730]: Ah found it!

File: tests/contrib/test_appengine.py [hunk 7]
```diff
@@ -610,11 +727,21 @@ def test_required(self):
...
         self.should_raise = False
         self.assertEqual(None, self.decorator.credentials)
 
+        # Access token refresh error should start the dance again
+        self.should_raise = AccessTokenRefreshError()
+        response = self.app.get('/foo_path')
+        self.should_raise = False
+        self.assertTrue(response.status.startswith('302'))
+        q = urllib.parse.parse_qs(
+            response.headers['Location'].split('?', 1)[1])
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:741]: `q` must be joking me with these variable names

File: tests/contrib/test_appengine.py [hunk 8]
```diff
@@ -627,6 +754,8 @@ def test_required(self):
         self.assertEqual('http://localhost/oauth2callback',
                          q['redirect_uri'][0])
 
+
+
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:756]: 3 newlines now exist where 1 is needed (ducks)

File: tests/contrib/test_appengine.py [hunk 9]
```diff
@@ -769,6 +898,38 @@ def test_decorator_from_client_secrets(self):
         self.assertEqual(self.decorator._revoke_uri,
                          self.decorator.credentials.revoke_uri)
 
+
```
Comment (dhermes) [inline tests/contrib/test_appengine.py]: Ditto on spurious newline! I am lint, hear me complain!

File: tests/contrib/test_appengine.py [hunk 10]
```diff
@@ -769,6 +898,38 @@ def test_decorator_from_client_secrets(self):
...
+    def test_decorator_from_client_secrets_toplevel(self):
+        decorator_patch = mock.patch(
+            'oauth2client.contrib.appengine.OAuth2DecoratorFromClientSecrets')
+
+        with decorator_patch as decorator_mock:
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:903]: Does `decorator_patch.__enter__` not return `self`?

File: tests/contrib/test_appengine.py [hunk 11]
```diff
@@ -769,6 +898,38 @@ def test_decorator_from_client_secrets(self):
...
+                filename,
+                'foo_scope',
+                cache=None,
+                message=None)
+
+    def test_decorator_from_client_secrets_bad_type(self):
+        # Note: this code path is actually not currently reachable, as the
+        # only types that oauth2client.clientsecrets can load is web and
+        # installed, so I have to forcably excute this code path. This should
+        # remain in case future kinds are added.
```
Comment (dhermes) [inline tests/contrib/test_appengine.py]: JINX

File: tests/contrib/test_appengine.py [hunk 12]
```diff
@@ -769,6 +898,38 @@ def test_decorator_from_client_secrets(self):
...
+            decorator_mock.assert_called_once_with(
+                filename,
+                'foo_scope',
+                cache=None,
+                message=None)
+
+    def test_decorator_from_client_secrets_bad_type(self):
+        # Note: this code path is actually not currently reachable, as the
+        # only types that oauth2client.clientsecrets can load is web and
+        # installed, so I have to forcably excute this code path. This should
```
Comment (dhermes) [inline tests/contrib/test_appengine.py]: `I`? forcibly?

**Comment by dhermes** [comment]
Finished review. Close to LGTM but no cigar.

File: tests/contrib/test_appengine.py (hunk 1 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:52]: Oy indeed.

File: tests/contrib/test_appengine.py (hunk 2 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:335]: Not that I know of?

File: tests/contrib/test_appengine.py (hunk 10 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:903]: I have no idea. Does it?

File: tests/contrib/test_appengine.py (hunk 3 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py]: I hate unittest so much. Fixed in incoming change.

File: tests/contrib/test_appengine.py (hunk 12 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py]: Fixed

File: tests/contrib/test_appengine.py (hunk 11 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py]: Fixed.

File: tests/contrib/test_appengine.py (hunk 9 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py]: Fixed.

File: tests/contrib/test_appengine.py (hunk 8 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:756]: Fixed.

File: tests/contrib/test_appengine.py (hunk 7 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:741]: Fixed. Sorry, thought I was a go programmer for a second.

File: tests/contrib/test_appengine.py (hunk 5 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py]: Fixed. As far as filing a bug, this entire module probably needs to be re-thought. This is minor compared to all of the other nonsense going on in here.

File: tests/contrib/test_appengine.py (hunk 10 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:903]: It does not.

File: tests/contrib/test_appengine.py (hunk 2 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:335]: Ah, there are tests for the ndb versions, but not the db versions.

**Comment by jonparrott** [comment]
Everything has been addressed, should be cigar now.

File: tests/contrib/test_appengine.py [hunk 13]
```diff
@@ -339,6 +417,27 @@ def setUp(self):
     def tearDown(self):
         self.testbed.deactivate()
 
+    def test_bad_ctor(self):
+        with self.assertRaises(ValueError):
+            StorageByKeyName(CredentialsModel, None, None)
```
Comment (dhermes) [inline tests/contrib/test_appengine.py:422]: Why you hate `unittest`? I think this is nice?

**Comment by dhermes** [comment]
LGTM

Pretty please file an issue to encapsulate some of the issues / notes you're referring to in the module under test.

File: tests/contrib/test_appengine.py (hunk 13 above)
Comment (theacodes) [inline tests/contrib/test_appengine.py:422]: I'm a `py.test` believer, and in `py.test` the context manager is the one true way of doing this, so there's no ambiguity. (The other way is only there for 2.4 support).

**Pull request merged**